File: app/src/main/java/com/nikola/jakshic/dagger/viewModel/HeroViewModel.java
package com.nikola.jakshic.dagger.viewModel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.arch.paging.PagedList;

import com.nikola.jakshic.dagger.Status;
import com.nikola.jakshic.dagger.model.Hero;
import com.nikola.jakshic.dagger.repository.HeroRepository;

import javax.inject.Inject;

public class HeroViewModel extends ViewModel {

    private static final String LOG_TAG = HeroViewModel.class.getSimpleName();

    private LiveData<PagedList<Hero>> list;
    private HeroRepository repository;
    private MutableLiveData<Status> status;
    private boolean initialFetch;

    @Inject
    public HeroViewModel(HeroRepository repository) {
        this.repository = repository;
        status = new MutableLiveData<>();
        status.setValue(Status.LOADING);
    }

    public void initialFetch(long id) {
        if (!initialFetch) {
            list = repository.fetchByGames(id);
            repository.fetchHeroes(status, id);
            initialFetch = true;
        }
    }

    public void fetchHeroes(long id) {
        repository.fetchHeroes(status, id);
    }

    public void sortByGames(long id){
        list = repository.fetchByGames(id);
    }

    public void sortByWinrate(long id){
        list = repository.fetchByWinrate(id);
    }

    public void sortByWins(long id){
        list = repository.fetchByWins(id);
    }

    public void sortByLosses(long id){
        list = repository.fetchByLosses(id);
    }

    public LiveData<PagedList<Hero>> getHeroes() {
        return list;
    }

    public MutableLiveData<Status> getStatus() {
        return status;
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/ui/profile/hero/HeroViewModel.kt
package com.nikola.jakshic.dagger.ui.profile.hero

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.nikola.jakshic.dagger.Status
import com.nikola.jakshic.dagger.model.Hero
import com.nikola.jakshic.dagger.repository.HeroRepository
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

class HeroViewModel @Inject constructor(private val repository: HeroRepository) : ViewModel() {

    lateinit var list: LiveData<List<Hero>>
        private set
    val status = MutableLiveData<Status>()
    private val disposables = CompositeDisposable()
    private var initialFetch = false

    fun initialFetch(id: Long){
        if (!initialFetch){
            initialFetch= true
            fetchHeroes(id)
            sortByGames(id)
        }
    }

    fun fetchHeroes(id: Long) {
        disposables.add(repository.fetchHeroes(status, id))
    }

    fun sortByGames(id: Long) {
        list = repository.sortByGames(id)
    }

    fun sortByWinRate(id: Long) {
        list = repository.sortByWinRate(id)
    }

    fun sortByWins(id: Long) {
        list = repository.sortByWins(id)
    }

    fun sortByLosses(id: Long) {
        list = repository.sortByLosses(id)
    }

    override fun onCleared() {
        disposables.clear()
    }
}